op-batcher/batcher/driver.go

```diff
@@ -158,19 +154,27 @@ func (l *BatchSubmitter) StartBatchSubmitting() error {
 	}
 
 	receiptsCh := make(chan txmgr.TxReceipt[txRef])
-	receiptsLoopCtx, cancelReceiptsLoopCtx := context.WithCancel(context.Background())
-	throttlingLoopCtx, cancelThrottlingLoopCtx := context.WithCancel(context.Background())
+
+	l.txpoolMutex.Lock()
+	l.txpoolState = TxpoolGood
+	l.txpoolMutex.Unlock()
+
+	// Channels used to signal between the loops
+	pendingBytesUpdated := make(chan int64)
+	blocksLoaded := make(chan struct{})
 
 	// DA throttling loop should always be started except for testing (indicated by ThrottleInterval == 0)
 	if l.Config.ThrottleInterval > 0 {
 		l.wg.Add(1)
-		go l.throttlingLoop(throttlingLoopCtx)
+		go l.throttlingLoop(l.wg, pendingBytesUpdated) // ranges over pendingBytesUpdated channel
 	} else {
 		l.Log.Warn("Throttling loop is DISABLED due to 0 throttle-interval. This should not be disabled in prod.")
 	}
-	l.wg.Add(2)
-	go l.processReceiptsLoop(receiptsLoopCtx, receiptsCh)                                    // receives from receiptsCh
-	go l.mainLoop(l.shutdownCtx, receiptsCh, cancelReceiptsLoopCtx, cancelThrottlingLoopCtx) // sends on receiptsCh
+
+	l.wg.Add(3)
+	go l.receiptsLoop(l.wg, receiptsCh)                                   // ranges over receiptsCh channel
+	go l.writeLoop(l.shutdownCtx, l.wg, receiptsCh, blocksLoaded)         // ranges over blocksLoaded, sends on receiptsCh + closes it when done
```

Comment:
(from call with @sebastianst)

The write loop does not actually directly send on the receiptsCh. It is responsible for spawning _multiple senders_ on that channel. Therefore there is a risk that we try to send on a closed channel (panic). We should check that we don't close that channel before all producers are done (I think we do via `txQueue.Wait()`.

Reply:
We should reflect this in the diagram in the readme. And code comments should be made more accurate.